src/testing/systest/supervisor/logged_process.zig

```diff
@@ -0,0 +1,329 @@
+    allocator.destroy(self);
+}
+
+pub fn state(self: *Self) State {
+    return self.current_state.load(.seq_cst);
+}
+
+pub fn start(
+    self: *Self,
+) !void {
+    self.expect_state_in(.{ .initial, .terminated, .completed });
+    defer self.expect_state_in(.{.running});
+
+    assert(self.child == null);
+
+    var child = std.process.Child.init(self.argv, self.allocator);
+
+    child.cwd = self.cwd;
+    child.env_map = self.options.env;
+    child.stdin_behavior = .Pipe;
+    child.stdout_behavior = .Ignore;
+    child.stderr_behavior = .Pipe;
+
+    try child.spawn();
+
+    errdefer {
+        _ = child.kill() catch {};
+    }
+
+    // Zig doesn't have non-blocking version of child.wait, so we use `BrokenPipe`
+    // on writing to child's stdin to detect if a child is dead in a non-blocking
+    // manner. Checks once a second second in a separate thread.
+    _ = try std.posix.fcntl(
+        child.stdin.?.handle,
+        std.posix.F.SETFL,
+        @as(u32, @bitCast(std.posix.O{ .NONBLOCK = true })),
+    );
+    self.stdin_thread = try std.Thread.spawn(
+        .{},
+        struct {
+            fn poll_broken_pipe(stdin: std.fs.File, process: *Self) void {
+                while (process.state() == .initial or process.state() == .running) {
+                    std.time.sleep(1 * std.time.ns_per_s);
+                    _ = stdin.write(&.{1}) catch |err| {
+                        switch (err) {
+                            error.WouldBlock => {}, // still running
+                            error.BrokenPipe,
+                            error.NotOpenForWriting,
+                            => {
+                                process.current_state.store(.completed, .seq_cst);
+                                break;
+                            },
+                            else => @panic(@errorName(err)),
+                        }
+                    };
+                }
+            }
+        }.poll_broken_pipe,
+        .{ child.stdin.?, self },
+    );
+
+    // The child process' stderr is echoed to stderr with a name prefix
+    self.stderr_thread = try std.Thread.spawn(
+        .{},
+        struct {
+            fn log_stderr(stderr: std.fs.File, process: *Self) void {
+                while (process.state() == .initial or process.state() == .running) {
+                    var buf: [1024 * 8]u8 = undefined;
+                    const line_opt = stderr.reader().readUntilDelimiterOrEof(
+                        &buf,
+                        '\n',
+                    ) catch |err| {
+                        log.warn("{s}: failed reading stderr: {any}", .{ process.name, err });
+                        continue;
+                    };
+                    if (line_opt) |line| {
+                        log.info("{s}: {s}", .{ process.name, line });
+                    } else {
+                        break;
+                    }
+                }
+            }
+        }.log_stderr,
```

Comment:
Good point about stderr not having a global lock, but I still don't feel good about this code. I must confess, at this point, this is 80% gut feeling, and I am totally ready to eventually admit that I was wrong here, but just not right away :-) 

I think we need to do _something_ with logs and error streams, but I am very not sure that a separate stderr reading thread is the correct something _here_, and I fear that implementing a wrong solution would disincentivize us from finding the right, simple solution here. 

So, let's "do nothing" here, let's just inherit stderr stream, make a mental note to "fix this later", and revisit it later. 

I have a suspicion that this _might_ actually be fine, provided that the child processes obey three rules:

* Each process log is self-identifying 
* Each log message fits into 4k limit of kernel's atomic writes
* Each log message is a result of a single `.write` syscall

If all the above holds, then I thing the log would _not_ be messed up. And, given that we write the child processes, we have the power to make the above points true, and they _seems_ like good ideas nonetheless.

Reply:
Hm, I'd maybe avoid parsing the logs. This is again, gut feeling, but for cfo, things definitely improved once we stopped trying to collect the logs, and instead went for binary "does it exit with zero". I think it makes more sense to treat this as a black box --- there's a certain contract that the cluster should fulfill, and it doesn't _really_ matter whether replica logs something or not.